jupyter_client/runapp.py

```diff
@@ -0,0 +1,157 @@
+        self.init_kernel_info()
+
+    def handle_sigint(self, *args):
+        if self.kernel_manager:
+            self.kernel_manager.interrupt_kernel()
+        else:
+            print("", file=sys.stderr)
+            error("Cannot interrupt kernels we didn't start.\n")
+
+    def init_kernel_info(self):
+        """Wait for a kernel to be ready, and store kernel info"""
+        timeout = self.kernel_timeout
+        tic = time.time()
+        self.kernel_client.hb_channel.unpause()
+        msg_id = self.kernel_client.kernel_info()
+        while True:
+            try:
+                reply = self.kernel_client.get_shell_msg(timeout=1)
+            except queue.Empty:
+                if (time.time() - tic) > timeout:
+                    raise RuntimeError("Kernel didn't respond to kernel_info_request")
+            else:
+                if reply['parent_header'].get('msg_id') == msg_id:
+                    self.kernel_info = reply['content']
+                    return
+
+    def start(self):
+        self.log.debug("jupyter run: starting...")
+        super(RunApp, self).start()
+        for filename in self.filenames_to_run:
+            self.log.debug("jupyter run: executing `%s`" % filename)
+            with open(filename) as fp:
+                cell = fp.read()
+            return_code = self.execute_printing_output(cell)
+            if return_code:
+                raise Exception("jupyter-run error running '%s'" % filename)
+
+    def execute_printing_output(self, cell):
+        """
+        Run a cell on a KernelClient
+        Any output from the cell will be displayed.
+        """
+        msg_id = self.kernel_client.execute(cell)
+        return_code = 0
+        while True:
+            try:
+                msg = self.kernel_client.get_iopub_msg(timeout=OUTPUT_TIMEOUT)
+            except queue.Empty:
+                raise TimeoutError("Timeout waiting for kernel output")
+    
+            if msg['parent_header'].get('msg_id') != msg_id:
+                continue
+            msg_type = msg['header']['msg_type']
+            content = msg['content']
+            if msg_type == 'status':
+                if content['execution_state'] == 'idle':
+                    # idle means output is done
+                    break
+            elif msg_type == 'stream':
+                stream = getattr(sys, content['name'])
+                stream.write(content['text'])
+                # TODO: remove this when all kernels use error msg_type:
+                if content['name'] == "stderr":
+                    return_code = 1
+            elif msg_type in ('display_data', 'execute_result', 'error'):
+                if msg_type == 'error':
+                    print('\n'.join(content['traceback']), file=sys.stderr)
+                    return_code = 1
+                else:
+                    sys.stdout.write(content['data'].get('text/plain', ''))
+            elif msg_type == 'aborted':
```

Comment:
This is not a message type. We need to get a message from the shell channel with `self.kernel_client.get_shell_msg`, check that it's an `execute_reply` message, and check its 'status' field.

http://jupyter-client.readthedocs.io/en/latest/messaging.html#execution-results